chain/chain/src/store/latest_witnesses.rs

```diff
@@ -0,0 +1,206 @@
+//! This module is responsible for storing the latest ChunkStateWitnesses observed
+//! by the node. The latest witnesses are stored in the database and can be fetched
+//! for analysis and debugging.
+//! The number of stored witnesses is limited. When the limit is reached
+//! the witness with the oldest height is removed from the database.
+//! At the moment this module is used only for debugging purposes.
+
+use std::io::ErrorKind;
+
+use near_primitives::hash::CryptoHash;
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_primitives::types::EpochId;
+use near_store::DBCol;
+
+use crate::ChainStoreAccess;
+
+use super::ChainStore;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_store::db::LATEST_WITNESSES_INFO;
+use rand::rngs::OsRng;
+use rand::RngCore;
+
+/// Maximum size of the latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_SIZE: ByteSize = ByteSize::gb(4);
+
+/// Maximum size of a single latest witness stored in the database.
+const SINGLE_LATEST_WITNESS_MAX_SIZE: ByteSize = ByteSize::mb(128);
+
+/// Maximum number of latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_COUNT: u64 = 60 * 30;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LatestWitnessesKey {
+    pub height: u64,
+    pub shard_id: u64,
+    pub epoch_id: EpochId,
+    /// Each witness has a random UUID to ensure that the key is unique.
+    /// It allows to store multiple witnesses with the same height and shard_id.
+    pub random_uuid: [u8; 16],
+}
+
+impl LatestWitnessesKey {
+    /// `LatestWitnessesKey` has custom serialization to ensure that the binary representation
+    /// starts with big-endian height and shard_id.
+    /// This allows to query using a key prefix to find all witnesses for a given height (and shard_id).
+    pub fn serialized(&self) -> [u8; 64] {
```

Comment:
why do we need manual ser/deser instead of just using borsh?
from what I understand we use borsh whenever possible when encoding data in the db

Reply:
I want the rows to be ordered by (height, shard_id) in big endian representation. I'm not sure if borsh would do that, it was quicker and easier to manually serialize it.